clang-tidy: Major `clang-tidy` fixes

LecrisUT opened the pull request:
# Manually fix various `clang-tidy` fixes

This PR concerns the commit 3a01389e46d53568d91d56f8748def93592a9f0f. This should be reviewed carefully. If I have the energy I might split them off at some point.

Basically what this commit does is:
- Moved initializers to outside of if statement 0725ad1a57f905b71f9b23327754a1d3481cfe9e
  (Prone to initialization errors)
- Moved loop variables initialization inside
- Removed conditional initializations
- Made a few static const arrays const
- Reduced goto statements
- Added missing assertion checks
- Resolved some memory leaks

I'll pull up some examples in the commit to show some of the changes

LecrisUT committed:
Add basic .clang-tidy file

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
[clang-tidy] Remove redundant const parameter

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
[clang-tidy] Align the structs

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
[clang-tidy] Redundant else after return

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
add style choice clang-tidy

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
[clang-tidy] Make static const variables const

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT commented on src/arithmetic.c:
```diff
@@ -132,15 +132,13 @@ static const char arithmetic_crystal_class_symbols[74][7] = {
-
     if (spgroup_number < 1 || spgroup_number > 230) {
         return 0;
     }
 
-    arth_number = arithmetic_crystal_classes[spgroup_number];
+    int arth_number = arithmetic_crystal_classes[spgroup_number];
     memcpy(symbol, arithmetic_crystal_class_symbols[arth_number], 7);
-    for (i = 0; i < 6; i++) {
+    for (int i = 0; i < 6; i++) {
```
This change moves the initializations near where the variables are used, e.g. only within the for loop.
- This allows the compiler to unroll the loops efficiently
- Also guarantees that the compiled nested functions have their own local variables and are thus more efficient in their memory access

atztogo replied:
It was necessary to support an old standard compiler. Now OK, I believe. If I remember correctly, I wasn't allowed even using `pow` function for that compiler...

LecrisUT commented on src/cell.c:
```diff
@@ -61,15 +61,14 @@ static int *get_overlap_table(const VecDBL *position, int cell_size,
-    if ((cell = (Cell *)malloc(sizeof(Cell))) == NULL) {
+    Cell *cell = (Cell *)malloc(sizeof(Cell));
+
+    if (cell == NULL) {
```
Here's an example of
- avoiding double instantiation
- report from `clang-tidy` discourages usage of variable instantiation in  the if loop. If an error occurs in the function call, the return value can be ill-defined

LecrisUT commented:
Main errors right now appear to be with the address sanitizer, which is to be expected with these many changes. Yep, some mistakes on my part.

Hmm, now the pytests are failing :thinking: I would have thought that these would pass if the ctests were passing.

LecrisUT committed:
[clang-tidy] Temporarily disable complexity check

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

atztogo commented:
> Hmm, now the pytests are failing 🤔 I would have thought that these would pass if the ctests were passing.

There are many naively distorted crystal structures with respect to ideal symmetry. ctests covers very basis ones, but pytest does more cases. It results in entering specific routines to recover from detailed different symptoms.

I feel this PR is dangerous. It would be helpful to divide into small commits.

LecrisUT commented:
> I feel this PR is dangerous. It would be helpful to divide into small commits.

Indeed that would be ideal, but it will take a bit of time yo detangle them. At least the safe ones we have it in the other 2 commits.

> There are many naively distorted crystal structures with respect to ideal symmetry. ctests covers very basis ones, but pytest does more cases. It results in entering specific routines to recover from detailed different symptoms.

Hmm I think we should address #291 and its equivalent googletest to actually report the specifics of how it fails. It should be possible to map the pytests 1-to-1 onto the ctests, but it would require a bit of organization, mainly split the unit-tests from the functional tests. All of the python and fortran api tests are only functional ones, and those should be identically covered and use a common set of inputs that we synchronize through a file. For this testpytests should be moved to `/tests` folder which we then split logically as well.

<details><summary>When to label a test as a unit-test or functional test</summary>

 ### unit test
- runs a single self-contained internal function, e.g. constructing and destructing the lattice object
- it does not touch any physical aspects
- used to debug compilation, segmentation issues etc.
- the main thing to investigate in codecov

### functional test
- runs the full calculation as the user would be doing
- requires input data
- tests the physical aspects of the library

### integration test (yes, there's one more technically)
- middle step between unit and functional test. Runs multiple fumctions covered by unit tests or the higher level functions
- tests if the data is transfered correctly between the functions, mainly for segmentation faults and unexpected crashes in the manipulation of data

</details>

LecrisUT committed:
[clang-tidy] Move initializers outside if statement

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
[clang-tidy] Various fixes

- Moved loop variables initialization inside
- Removed conditional initializations
- Made a few static const arrays const
- Reduced goto statements
- Added missing assertion checks
- Resolved some memory leaks

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
[null-checks] Missing null-check

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>
